Begin UI Refactor with Responsiveness & New Menus

danfinlay opened the pull request:
Ready for QA! Needs thorough QA to make sure all features are still accessible, useful, and their formatting is not messed up.

Rather than let this `NewUI` branch run any longer, we've made changes to it so it can be "production ready" asap.

This PR represents our first suspicion that *this may be ready to go to production*!

Also, it looks good in different view sizes, from desktop tabs to mobile.

Also includes the new work by Kyokan Labs, where they implemented the new drop-down designs by Christian, so this is starting to inch towards our new design template.

I'll keep track of outstanding changes before we can merge this here:

- [x] Custom RPC items remain checked and clicking them is ineffective.
- [x] Selecting `Localhost 8545` ineffective.

danfinlay committed:
Began creating new UI template

danfinlay committed:
Merge branch 'master' into NewUI

danfinlay committed:
Make folder for responsive UI

danfinlay committed:
Get duplicate UI template working

danfinlay committed:
Add responsive UI dev guide

danfinlay committed:
Make responsive UI more flexy

danfinlay committed:
Merge branch 'master' into NewUI

sdtsui committed:
Implement redesigned dropdown

sdtsui committed:
Add tests for new dropdown component

sdtsui committed:
Brighten dropdown menu item\'s text

kumavis committed:
Merge pull request #1772 from sdtsui/new-dropdown

Implement NewUI Dropdown Component

sdtsui committed:
Cleanup

Fix lint error breaking gulp build

Add presentational options menus

sdtsui committed:
Fix click handlers on AccountOptionsMenus

sdtsui committed:
Bump version of menu-droppo to include bugfix

sdtsui committed:
Add new tests for dropdown component

sdtsui committed:
Remove accounts screen

sdtsui committed:
Remove account options icons

sdtsui committed:
Hook up new dropdown components

sdtsui committed:
Reorder Account Selector and Account Options

FlySwatter committed:
Merge pull request #1790 from sdtsui/newui-account-dropdowns

[newui] - Implement Advanced Option Menu & Account Switching Menu

danfinlay committed:
Move responsive ui into its own folder for easier merges

danfinlay committed:
Merge branch 'master' into NewUI

danfinlay committed:
Fix css links

danfinlay committed:
Fix reference

FlySwatter committed:
Merge pull request #1802 from MetaMask/RestructureNewUI

Restructure new ui branch folders

danfinlay committed:
Make main account detail view more responsive

danfinlay committed:
Prevent buy button from floating

danfinlay committed:
Improve account-detail panel responsiveness

sdtsui committed:
Add cursor to account menu buttons

kumavis committed:
Merge pull request #1815 from sdtsui/fix/menu-cursor

Add cursor to account menu buttons

danfinlay committed:
Css tweaks

danfinlay committed:
Merge branch 'NewUI' of github.com:MetaMask/metamask-plugin into NewUI

danfinlay committed:
Replace ui with responsive-ui

danfinlay committed:
Restore sort order fix

danfinlay committed:
Remove responsive folder from gulp

danfinlay committed:
Lots of flex rearrangement on account detail view

Includes removal of ReactCssTransitionGroup for a simpler UI refactor.

sdtsui committed:
Bump version of menu-droppo, no longer needs css transitions

sdtsui committed:
Remove Account Settings item from optionsMenu, unnecessary

sdtsui committed:
Make account selection dropdown menu scrollable when too large for view

sdtsui committed:
Re-center network dropdown

FlySwatter committed:
Merge pull request #1831 from sdtsui/pre-merge-fixes

Add minor fixes, including account selection scroll

danfinlay committed:
Set initial scale for mobile.

danfinlay committed:
Adjust mobile scale for smaller devices

danfinlay committed:
Remove object spread syntax

danfinlay committed:
Fix custom provider indication

flyswatter commented:
Okay, fixed the one bug I was aware of, @tmashuang this is ready for a QA pass when you have time.

sdtsui commented on docs/responsive-ui-dev.md:
```diff
@@ -0,0 +1,11 @@
+# Developing our Responsive UI
```
Are these instructions still applicable now that the folders are merged?  Suggesting that we remove, or add some context about responsive development instead.

sdtsui commented on package.json:
```diff
@@ -87,7 +87,7 @@
     "jazzicon": "^1.2.0",
     "loglevel": "^1.4.1",
-    "menu-droppo": "^1.1.0",
+    "menu-droppo": "2.0.0",
```
Might be low priority: could this be `2.0.1` since it's merged in `menu-droppo`?

sdtsui commented on ui/app/add-token.js:
```diff
@@ -86,7 +86,7 @@ AddTokenScreen.prototype.render = function () {
             h('span', {
               style: { fontWeight: 'bold', paddingRight: '10px'},
-            }, 'Token Symbol'),
+            }, 'Token Sybmol'),
```
typo?  (s*ym*bol..)

sdtsui commented on ui/app/account-detail.js:
```diff
@@ -54,12 +50,13 @@ AccountDetailScreen.prototype.render = function () {
   return (
 
-    h('.account-detail-section', [
+    h('.account-detail-section.full-flex-height', [
```
:+1:

sdtsui reviewed (commented):
Minor suggestions.

danfinlay committed:
Bump menu-droppo to 2.0.1

danfinlay committed:
Remove deprecated responsive dev docs

danfinlay committed:
Fix spelling

flyswatter commented:
Good catches, @sdtsui, implemented all those suggestions.

tmashuang commented:
- [ ] In info.js, the Twitter link should be on a new line. Current CSS display is set to inline-block from fa, font-awesome, ```display: block``` should fix this for twitter link.
- [ ] Instance of Loading Spinner persisting after sending transactions. 
- [ ] Loading Spinner does not properly center
Note: When the extension URL is open in a window no popup appears when interaction with Dapps. Don't know if this is the intended functionality.

danfinlay committed:
Fix info page formatting

danfinlay committed:
Restore support link

danfinlay committed:
Remove unused parameter

danfinlay committed:
Fix loading indication placement

danfinlay committed:
Ensure loading indication is full screen

flyswatter commented:
@tmashuang Thanks for the QA, I fixed 2/4 of them. On the remaining issues:

> Instance of Loading Spinner persisting after sending transactions.

I briefly saw this, it was related to a nonce not being an integer, which was a really weird bug, because we already adjusted provider engine to allow non-integer nonces. This then went away, and I haven't seen it again. Can you get this again? I think it may be fixed.

> Note: When the extension URL is open in a window no popup appears when interaction with Dapps. Don't know if this is the intended functionality.

This is because we have some logic that says "If the UI is already open when a TX is proposed, don't open another one". We'll have to revise that logic for Mascara, but for now, it's consistent with how the extension UI behaves correctly.

So anyways, if you could re-check that spinner-post-tx bug, and let me know if it's fixed, and if there's anything else, I think we're getting close.

danfinlay committed:
Fix viewport width to 1

danfinlay committed:
Disable user zoom in mobile mode

tmashuang commented on ui/app/components/pending-tx.js:
```diff
@@ -15,7 +15,7 @@ const addressSummary = util.addressSummary
 const nameForAddress = require('../../lib/contract-namer')
 const BNInput = require('./bn-as-decimal-input')
 
-const MIN_GAS_PRICE_GWEI_BN = new BN(1)
```
Are we considering going back a minimal gas price of 2? #1796

flyswatter replied:
No, sorry, when I copied the `responsive-ui` folder over `ui` the merge seems to be prioritizing the `NewUI` over `master`, so it's changing some things like this.  Good catch, I'll review the diffs for other things like this.

flyswatter replied:
Yeah I've read over the diffs again, I didn't see any other changes we missed.

danfinlay committed:
Set min gas price to 1 gwei

danfinlay committed:
Merge branch 'master' into NewUI

danfinlay committed:
Merge branch 'master' into NewUI

danfinlay committed:
Bump changelog

tmashuang commented on ui/app/components/token-list.js:
```diff
@@ -76,17 +78,35 @@ TokenList.prototype.render = function () {
+
+  let msg
+  if (tokens.length > 0) {
+    msg = `You own ${tokens.length} tokens`
```
Possibility for only 1 token would display "You own 1 tokens"

tmashuang commented:
Last one, in recover-seed/confirmation.js the width is for flex-space-between is rather large, 50%, could be reduced to 20.

tmashuang commented:
NewUI in FF
<img width="1212" alt="newui_in_ff" src="https://user-images.githubusercontent.com/13376180/28696566-3f983e02-72eb-11e7-84a6-c30efbecde70.png">

frankiebee commented:
I seem to be getting something slightly different in fullscreen in mascara am i missing some new build step?

<img width="1430" alt="screen shot 2017-07-27 at 7 34 07 pm" src="https://user-images.githubusercontent.com/11225267/28696745-90851c30-7305-11e7-9ca0-12b60e23bea8.png">

danfinlay committed:
Correct token pluralization for one token

danfinlay committed:
Limit max width of seed word conf screen

flyswatter commented:
@frankiebee That's how it's supposed to work. It's flexy. Try resizing the window. It's approaching the layout that Christian has passed us.

danfinlay committed:
Fix react warning

danfinlay committed:
Correct viewport param

flyswatter commented:
Added a max-width per @tmashuang's advice. https://github.com/MetaMask/metamask-extension/pull/1835/files#diff-05f3f3e6a668f4f66e3c66bd417dd669R27

flyswatter commented:
Interesting that Firefox automatically flexes the action-view wide, and Chrome doesn't. We might want to simply have hard-coded dimensions on the extension's html vs mascara or mobile.

flyswatter commented:
Firefox does not have good style inspecting of its WebExtension pop-ups. I can't even look at which CSS property is forcing the view wide like that. I've contacted them on IRC (which has been one of my most responsive channels with them in the past).

Options:
- Hard coding popup dimensions, using a 2nd html file for responsiveness
- Accepting that Firefox has strange dimensions when it's used

flyswatter commented:
Whoops didn't mean to close.

danfinlay committed:
Create distinct responsive 'home.html' file, hard-code popup.html size

Because firefox was having inconsistent sizing, made a second html file for forcing the view to a certain size. Still allows us to develop a responsive interface via the `home.html` file, which shares all the same react JS & CSS as popup.html.

flyswatter commented:
Okay, we now have two UI html files. `popup.html` includes inline styles to force it to the traditional size, like this:

![screen shot 2017-07-28 at 11 07 50 am](https://user-images.githubusercontent.com/542863/28730419-06035984-7385-11e7-8980-34feafa98c11.png)

The new `home.html` file is the one that has no fixed sizes, and supports a fully responsive interface.

sdtsui commented:
fyi for @flyswatter - I noticed in both popup screenshots (Thomas' firefox one, and above): we're losing a few px off the bottom of the account address. Worth fixing?
(Doesn't happen with Frankie's responsive screenshot.)

sdtsui commented:
Noticed this on latest. Global handler in `menu-droppo 2.0.1` appears to be having unintended side effects on dropdown behavior. Fixing now. 

![2 0 1 effects](https://user-images.githubusercontent.com/8230144/28737363-02e40ca8-73a3-11e7-9446-e1b5a06592c8.gif)

sdtsui committed:
Fix dropdown toggle behavior - account dropdowns

sdtsui committed:
Fix dropdown toggle behavior - settings

sdtsui committed:
Fix dropdown behavior - network

sdtsui commented:
@flyswatter @tmashuang - PRing on top of this one to fix an issue I spotted.
#1841

sdtsui commented:
I was preparing some docs for the state atom to make the redesign work easier.
Along the way, I was able to repro the post-tx-spinner issue in popup. Sharing here. 
```
651fec5 - (origin/NewUI, metamask/NewUI) Create distinct responsive 'home.html' file, hard-code popup.html size (8 hours ago) <Dan Finlay>hours ago) <Dan Finlay>
```

![post-tx-spinner](https://user-images.githubusercontent.com/8230144/28741186-cd1b02d4-73c5-11e7-929e-2e57cbb9a62d.gif)


Hope that helps.

sdtsui commented:
Adding a bit more context, from the state atom:
![](https://puu.sh/wVQYr/a3e8926592.png)

Full Gist:
https://gist.github.com/sdtsui/65a03c1f6eb4257bb2cdf16da4e1ddaa

flyswatter commented:
> we're losing a few px off the bottom of the account address. Worth fixing?

It's only on Firefox, so it's worth fixing for FF, but also, FF hasn't made it easy to analyze CSS for action-views like this, so it's hard to prioritize very seriously. They need to bring their extension debugging up a notch for anyone to bring really high quality extensions.

To reiterate something I said in our dev channel:

I think it'd be smart to deploy our current master before merging this, so we get more QA time on it. Our current master has some useful blacklist enhancements, while this style bump is not exactly urgent, and would benefit from more developer-usage time, we'll probably work through some kinks like that spinner bug.

I'm out for a week now, so good luck everyone on that bug and others!  Thanks for holding down the fort, good luck!

sdtsui commented:
:+1: , thanks @flyswatter .  Enjoy the week off!

@Zanibas - shoot me a message on Gitter when you get this, we can team on next steps early Monday.  It'd be helpful to understand more about the deployment process as well.

sdtsui commented:
![tx-spinner-issue-on-newui-with-txdata](https://user-images.githubusercontent.com/8230144/28798071-3912da32-75f8-11e7-98f9-b4adda797fb5.gif)

sdtsui commented:
![tx-spinner-issue-on-newui-with-txdata-2](https://user-images.githubusercontent.com/8230144/28798077-3f151760-75f8-11e7-8516-2df4e931c391.gif)

sdtsui commented:
tx repro gifs we recorded live:  cc for @Zanibas

2-am-zzz committed:
Merge pull request #1841 from sdtsui/fix/dropdown-toggle

[Part-1] Fix/dropdown toggle

kumavis commented:
failing on lint - theres some unwated semicolons in there

kumavis committed:
lint fix

kumavis commented:
I think we lost a couple features in this PR
1) QR code display
2) copy-to-clipboard support

kumavis commented:
I think we lost the QR code display in this PR

@sdtsui @Zanibas

frankiebee commented:
when the window is at 797x781 the history and tokens snap to different sides
<img width="802" alt="screen shot 2017-08-01 at 5 49 16 pm" src="https://user-images.githubusercontent.com/11225267/28848674-df41179a-76e1-11e7-8e81-b4f0a9f0a42e.png">
<img width="815" alt="screen shot 2017-08-01 at 5 49 22 pm" src="https://user-images.githubusercontent.com/11225267/28848675-df49f874-76e1-11e7-9be0-80254aa99cb1.png">

kumavis commented:
@frankiebee thats intentional, its working towards the new design
it should prolly start at a thinner width though

kumavis commented:
a lot of the targets feel really small on mobile (dropdown caret/dots, dropdown options)

sdtsui commented:
Thanks @frankiebee @kumavis, we'll batch with the design call and review tomorrow. And thanks for the lint fix!

I wasn't aware QR code was staying: impression I got from Dan was to implement prototypes and remove all of the old buttons.  I'll add back with some sensible text for the option, but we should probably sync on any expected functionality losses for `newui` before I get farther along with it.

For example, newui doesn't have a QR code screen or a design for the 'custom RPC' option in network drop-down.

My instructions were to skip send token screen as well, and leave out anything without screens. Thoughts?

Finally, do either of you have thoughts on the tx spinner issue? Aside from some minor tweaks, that appears to be the main thing holding this PR back. Cheers.

Commit:
Merge branch 'NewUI' of github.com:MetaMask/metamask-plugin into NewUI

Zanibas commented:
Exporting a private key needs to be fixed (overflowing off-screen)

sdtsui commented:
- Synced on above in design call. Thanks for clarifying, @kumavis .
- The issue raised by @tmashuang has been reproduced in #1853 .

Commit:
Add QR functionality

Zanibas commented:
Some scrollbars are not wide enough. See notices and their differing sidebars.
The sizing is inconsistent at first-time

Commit:
Remove 100% properties that were messing up sizing.

Commit:
Add maximum width for private key reveal

Commit:
Fix wonky widths for notices.

Commit:
Fix fox positioning.

Zanibas commented:
We're going to stay with the text sizing on the droppos for now until later--not a blocking issue. We'll fix before anyone touches mobile, I promise :wink: 

We have to wait for the adjustment to the new font anyways to make semi-permanent calls on this.

kumavis commented:
i wouldnt consider it blocking - but i do frequently use mascara on mobile, which is currently using the new ui

Zanibas commented:
Okay, we'll elevate them then scale down if necessary.

sdtsui committed:
Adjust account checksumAddress location up

sdtsui committed:
Disable account selection dropdown from account details

sdtsui committed:
Move accountselector menu-droppo up to app-header

sdtsui committed:
Position account switcher icon back in app header

sdtsui committed:
Add note with previous fa-angle-down for future refactor

sdtsui committed:
De-dupe click handler for sandwich-expando

sdtsui committed:
Increase size of network dropdown

sdtsui committed:
Increase size of settings dropdown and account settings dropdown

sdtsui committed:
Increase size of accountSelection dropdown

sdtsui committed:
Allow new accounts selector to handle clicks

sdtsui committed:
Lint ui/app

sdtsui committed:
Adjust top and bottom padding of accountSwitcher

sdtsui committed:
Adjust padding of accountSwitcher dropdown

kumavis committed:
Merge pull request #1858 from sdtsui/NewUI

UI Tweaks, Minor Bugfixes, and Dropdown Resizing

kumavis committed:
Merge branch 'master' into NewUI

kumavis commented:
doesn't like the spread operator
https://github.com/MetaMask/metamask-extension/blob/5a9257208c858d11b94659518ef2375384aeac46/ui/app/components/dropdown.js#L76

kumavis commented:
specifically `nyc`'s `istanbul` (coverage) does not like it

kumavis commented:
i just noticed we had a [silent ci failure](https://circleci.com/gh/MetaMask/metamask-extension/2682) at 24a13d116ce4

```
Integration build failure { Error: ENOENT: no such file or directory, unlink '/home/ubuntu/metamask-extension/test/integration/bundle.js'
    at Object.fs.unlinkSync (fs.js:1066:18)
    at Object.<anonymous> (/home/ubuntu/metamask-extension/test/integration/index.js:11:6)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Function.Module.runMain (module.js:605:10)
    at startup (bootstrap_node.js:158:16)
    at bootstrap_node.js:575:3
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/ubuntu/metamask-extension/test/integration/bundle.js' }
```

kumavis committed:
ui - dropdown - use Object.assign syntax to appease coverage parser

kumavis committed:
test - integration - throw error on failure

kumavis commented:
ok now it doesnt fail silently -- goood

kumavis committed:
test - integration - remove unused code

kumavis committed:
test - integration - syntax - prefer const

kumavis committed:
Merge branch 'master' into NewUI

kumavis committed:
test - integration - better error handling

kumavis committed:
Merge branch 'NewUI' of github.com:MetaMask/metamask-extension into NewUI

kumavis commented:
some sort of fatal integration test?

kumavis commented:
thats right - our integration tests either pass or hang

sdtsui committed:
Re-enable css transitions for dropdowns in header, needs menu-droppo library update

sdtsui committed:
Bring menu-droppo component into project, remove as a dependency

sdtsui commented:
@Zanibas : #1862 re-enables dropdown css transitions, as discussed this morning.  Please see that PR for details.

sdtsui committed:
Adds early breakpoint from @frankiebee 's + @kumavis 's CR

sdtsui committed:
Cleanup breakpoint css + comment on hackiness

kumavis committed:
Merge pull request #1863 from sdtsui/newui-early-break

Add early breakpoint from @frankiebee 's + @kumavis 's CR of #1835

kumavis committed:
Merge pull request #1862 from sdtsui/add-rct

Re-enable css transitions, only for dropdowns in app header

sdtsui commented:
#1863 adds early breakpoint and resolved snapping weirdness.

sdtsui commented:
@kumavis  thanks for merging. Looks like menu-droppo needs to be linted. Checking it out.

sdtsui committed:
Lint menu-droppo

sdtsui commented:
@kumavis /\

kumavis committed:
Merge pull request #1864 from sdtsui/lint-droppo

Lint menu-droppo

sdtsui commented:
.. /\ looks like more silent failures.
```
> metamask-crx@0.0.0 genStates /home/ubuntu/metamask-extension
> node development/genStates.js
```
It'll wait 10m and probably time out.

kumavis commented:
our integration tests either pass or hang on failure

sdtsui committed:
Fix integration test failures: unnecessary sandwich-expando event stopPropagation

sdtsui committed:
Fix integration test failures: ensure qr code is accessible

kumavis committed:
Merge pull request #1867 from sdtsui/NewUI

Fix integration tests for NewUI

kumavis commented:
@sdtsui you can push directly to NewUI if you like (invited you to the repo https://github.com/MetaMask/metamask-extension/invitations)

kumavis committed:
Merge branch 'master' into NewUI

kumavis commented:
tests pass! awesome

kumavis commented:
will review and merge it all soon, maybe tomorrow

sdtsui reviewed (commented):
Cool, thanks - I'll respond to any changes on Monday.

sdtsui committed:
Improve readability of changelog

kumavis commented:
This is beautiful 👌
